src/data/complex/basic.lean

```diff
@@ -196,35 +196,41 @@ instance : star_ring ℂ :=
 /-! ### Norm squared -/
 
 /-- The norm squared function. -/
-@[pp_nodot] def norm_sq (z : ℂ) : ℝ := z.re * z.re + z.im * z.im
+@[pp_nodot] def norm_sq : monoid_with_zero_hom ℂ ℝ :=
+{ to_fun := λ z, z.re * z.re + z.im * z.im,
+  map_zero' := by simp,
+  map_one' := by simp,
+  map_mul' := λ z w, by { dsimp, ring } }
+
+lemma norm_sq_def (z : ℂ) : norm_sq z = z.re * z.re + z.im * z.im := rfl
 
 @[simp] lemma norm_sq_of_real (r : ℝ) : norm_sq r = r * r :=
 by simp [norm_sq]
 
-@[simp] lemma norm_sq_zero : norm_sq 0 = 0 := by simp [norm_sq]
-@[simp] lemma norm_sq_one : norm_sq 1 = 1 := by simp [norm_sq]
+lemma norm_sq_zero : norm_sq 0 = 0 := norm_sq.map_zero
+lemma norm_sq_one : norm_sq 1 = 1 := norm_sq.map_one
 @[simp] lemma norm_sq_I : norm_sq I = 1 := by simp [norm_sq]
 
 lemma norm_sq_nonneg (z : ℂ) : 0 ≤ norm_sq z :=
 add_nonneg (mul_self_nonneg _) (mul_self_nonneg _)
 
-@[simp] lemma norm_sq_eq_zero {z : ℂ} : norm_sq z = 0 ↔ z = 0 :=
+lemma norm_sq_eq_zero {z : ℂ} : norm_sq z = 0 ↔ z = 0 :=
```

Comment:
Is this lemma somewhere in `monoid_with_zero_hom` too?

Reply:
Yes but the proof of `complex.field` depends on `complex.norm_sq_eq_zero`, so we can't reuse the proof.

Reply:
`norm_sq.map_eq_zero` perhaps?

Reply:
Just giving a more explicit explanation for my own benefit: `monoid_with_zero_hom.map_eq_zero` requires the domain of the `monoid_with_zero_hom` to be a `group_with_zero`, but that instance for `ℂ` requires this result about `norm_sq`.

The proof used here relies instead on the fact that `ℝ` is a `linear_ordered_ring`.